src/client/metric.rs

```diff
@@ -0,0 +1,528 @@
+use byteorder::WriteBytesExt;
+use memmap::{Mmap, MmapViewSync, Protection};
+use std::ffi::CString;
+use std::io;
+use std::io::{Cursor, Write};
+use std::mem;
+
+const ITEM_BIT_LEN: usize = 10;
+
+pub const I32_METRIC_TYPE_CODE: u32 = 0;
+pub const U32_METRIC_TYPE_CODE: u32 = 1;
+pub const I64_METRIC_TYPE_CODE: u32 = 2;
+pub const U64_METRIC_TYPE_CODE: u32 = 3;
+pub const F32_METRIC_TYPE_CODE: u32 = 4;
+pub const F64_METRIC_TYPE_CODE: u32 = 5;
+pub const STRING_METRIC_TYPE_CODE: u32 = 6;
+
+/// Generic type for any Metric's value
+pub trait MetricType {
+    /// Returns the MMV metric type code
+    fn type_code(&self) -> u32;
+    /// Writes the byte representation of the value to a writer.
+    ///
+    /// For integer and float types, the byte sequence is little endian.
+    ///
+    /// For the string type, the UTF-8 byte sequence is suffixed with a null byte.
+    fn write_to_writer<W: WriteBytesExt>(&self, writer: &mut W)
+        -> io::Result<()>;
+}
+
+macro_rules! impl_metric_type_for (
+    ($typ:tt, $base_typ:tt, $type_code:expr) => (
+        impl MetricType for $typ {
+            
+            fn type_code(&self) -> u32 {
+                $type_code
+            }
+
+            fn write_to_writer<W: WriteBytesExt>(&self, mut w: &mut W)
+            -> io::Result<()> {
+                w.write_u64::<super::Endian>(
+                    unsafe {
+                        mem::transmute::<$typ, $base_typ>(*self) as u64
+                    }
+                )
+            }
+
+        }
+    )
+);
+
+impl_metric_type_for!(i32, u32, I32_METRIC_TYPE_CODE);
+impl_metric_type_for!(u32, u32, U32_METRIC_TYPE_CODE);
+impl_metric_type_for!(i64, u64, I64_METRIC_TYPE_CODE);
+impl_metric_type_for!(u64, u64, U64_METRIC_TYPE_CODE);
+impl_metric_type_for!(f32, u32, F32_METRIC_TYPE_CODE);
+impl_metric_type_for!(f64, u64, F64_METRIC_TYPE_CODE);
+
+impl MetricType for String {
+    fn type_code(&self) -> u32 {
+        STRING_METRIC_TYPE_CODE
+    }
+
+    fn write_to_writer<W: Write>(&self, mut writer: &mut W) -> io::Result<()> {
+        writer.write_all(CString::new(self.as_str())?.to_bytes_with_nul())
```

Comment:
It would probably be faster to do this without `CString` allocation, either formatted like `write!(writer, "{}\0", self)` or with separate calls `writer.write_all(self.as_bytes())?; writer.write_all(&[0])`.
(ditto in a few other places you're creating temporary `CString`s.)